File: implement-shell-tools/cat/cat.js
const fs = require("fs");

const args = process.argv.slice(2);

let flag = "";
let filePaths = [];

if (args[0] && args[0].startsWith("-")) {
  flag = args[0];
  filePaths = args.slice(1);
} else {
  filePaths = args;
}

// Read all files into one string
let content = filePaths.map((file) => fs.readFileSync(file, "utf-8")).join("");

// Split into lines
let lines = content.split(/\r?\n/);

// Remove the last empty line if the file ends with a newline
if (lines[lines.length - 1] === "") {
  lines.pop();
}

// Handle flag ==> -n adding number to the line;
if (flag === "-n") {
  lines = lines.map((line, index) => {
    return `${index + 1}  ${line}`;
  });
}

// Handle flag ==> -b avoids adding number to empty line
if (flag === "-b") {
  let lineNumber = 1;

  lines = lines.map((line) => {
    if (line.trim() === "") {
      return line;
    }

    return `${lineNumber++}  ${line}`;
  });
}

process.stdout.write(lines.join("\n"));

File: implement-shell-tools/cat/cat.py
# in built module
import argparse

parser = argparse.ArgumentParser()

parser.add_argument("-n", action="store_true")
parser.add_argument("-b", action="store_true")
parser.add_argument("file_paths", nargs="+")

args = parser.parse_args()

flag = ""

if args.n:
    flag = "-n"
elif args.b:
    flag = "-b"

file_paths = args.file_paths


# Read all files and combine their contents into one string
content = ""

for file in file_paths:
    with open(file, "r") as f:
        content += f.read()


# Split the file content into separate lines
lines = content.splitlines()


# Handle -n flag: add a number to every line
if flag == "-n":
    new_lines = []

    for index, line in enumerate(lines):
        new_lines.append(f"{index + 1}  {line}")

    lines = new_lines


# Handle -b flag: number only non-empty lines
if flag == "-b":
    line_number = 1
    new_lines = []

    for line in lines:
        # Keep empty lines without adding numbers
        if line.strip() == "":
            new_lines.append(line)

        # Add a number only to lines that contain text
        else:
            new_lines.append(f"{line_number}  {line}")
            line_number += 1

    lines = new_lines

print("\n".join(lines))
